app/src/main/java/com/example/chrono/main/circuit/CircuitTimerActivity.kt

```diff
@@ -0,0 +1,163 @@
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_circuit_timer)
+        bind = DataBindingUtil.setContentView(this, R.layout.activity_circuit_timer)
+
+        loadTimer(
+            GsonBuilder().create()
+                .fromJson(intent.getStringExtra("circuitObject"), CircuitObject::class.java)
+        )
+
+        // Initialize stuff perhaps
+
+        bind!!.startbutton.setOnClickListener {
+            workout()
+            timerState = TimerState.RUNNING
+            updateButtonUI()
+        }
+
+        bind!!.pausebutton.setOnClickListener {
+            countdown.cancel()
+            timerState = TimerState.PAUSED
+            updateButtonUI()
+        }
+
+        bind!!.resumebutton.setOnClickListener {
+            createTimer(secondsLeft.toInt(), true)
+            timerState = TimerState.RUNNING
+            updateButtonUI()
+        }
+
+        bind!!.closeButton.setOnClickListener {
+            // Make sure that the timer is shut down
+            countdown.cancel()
+            setResult(Activity.RESULT_CANCELED)
+            finish()
+        }
+    }
+
+    private fun createTimer(seconds: Int, wasPaused: Boolean) {
+        var time = seconds.toFloat() * 1000 + 250
+        if (wasPaused) {
+            time = secondsLeft * 1000 + 250
+        }
+        countdown = object : CountDownTimer(time.toLong(), 250) {
+            override fun onTick(p0: Long) {
+                if ((p0.toFloat().roundToInt() / 1000.0f) != secondsLeft) {
+                    secondsLeft = (p0.toFloat() / 1000.0f).roundToInt().toFloat()
+                    updateTimerUI()
+                }
+            }
+
+            override fun onFinish() {
+                if (currentSet != 0) {
+                    when (runningState) {
+                        RunningState.INIT -> {
+                            workout()
+                        }
+                        RunningState.WORK -> {
+                            rest()
+                        }
+                        RunningState.REST -> {
+                            workout()
+                        }
+                    }
+                } else {
+                    isDone()
+                }
+            }
+        }.start()
+    }
+
+    private fun isDone() {
+        Toast.makeText(this, "It's done", Toast.LENGTH_SHORT).show()
+        timerState = TimerState.INIT
+        runningState = RunningState.INIT
+        updateButtonUI()
+    }
+
+    // Update UI for every tick, possibly need to do more in the future
+    fun updateTimerUI() {
+        bind!!.countdown.text = (secondsLeft).toInt().toString()
+    }
+
+    private fun updateButtonUI() {
+        when (timerState) {
+            TimerState.INIT -> {
+                bind!!.initButtonLay.visibility = View.VISIBLE
+                bind!!.runButtonLay.visibility = View.GONE
+                bind!!.pauseButtonLay.visibility = View.GONE
+                bind!!.countdown.text = "0"
```

Comment:
Start text should be the set workout time.

Reply:
Depends on the design, hence this is temporary